sdk/database/newdbplugin/plugin_client.go

```diff
@@ -0,0 +1,80 @@
+package newdbplugin
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	log "github.com/hashicorp/go-hclog"
+	plugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/sdk/helper/pluginutil"
+)
+
+// DatabasePluginClient embeds a databasePluginRPCClient and wraps it's Close
+// method to also call Kill() on the plugin.Client.
+type DatabasePluginClient struct {
+	client *plugin.Client
+	sync.Mutex
+
+	Database
+}
+
+// This wraps the Close call and ensures we both close the database connection
+// and kill the plugin.
+func (dc *DatabasePluginClient) Close() error {
+	err := dc.Database.Close()
+	dc.client.Kill()
+
+	return err
+}
+
+// NewPluginClient returns a databaseRPCClient with a connection to a running
+// plugin. The client is wrapped in a DatabasePluginClient object to ensure the
+// plugin is killed on call of Close().
+func NewPluginClient(ctx context.Context, sys pluginutil.RunnerUtil, pluginRunner *pluginutil.PluginRunner, logger log.Logger, isMetadataMode bool) (Database, error) {
+	// pluginSets is the map of plugins we can dispense.
+	pluginSets := map[int]plugin.PluginSet{
+		5: plugin.PluginSet{
+			"database": new(GRPCDatabasePlugin),
+		},
+	}
```

Comment:
Not a dumb question at all. It does need to be able to connect to old plugins, however I decided to do that in a different way in order to keep the old and new packages separate and self-contained. The DB engine will attempt to load the plugin as a v5 plugin. If it fails it will fall back to loading it as a v4 plugin by calling the same functions but in the v4 package.

Reply:
Is that handled in a different PR (seems like it is from offline context)? It'd be great to look at the logic and see how it does the upgrade.

Reply:
This is handled in the next PR in the updated engine.